File: ch10/ch10_ws/reliability_test.py
import rclpy
from rclpy.node import Node
from nav_msgs.msg import Odometry
from rclpy import qos
from rclpy.qos import QoSProfile

class OdomPublisherSubscriber(Node):
    def __init__(self):
        super().__init__('odom_publisher_subscriber')

        qos_profile = qos.QoSProfile(depth=10,
                                     reliability=qos.ReliabilityPolicy.BEST_EFFORT,
                                     durability=qos.DurabilityPolicy.VOLATILE)
        
        self.odom_publisher_ = self.create_publisher(Odometry, 'odom', qos.qos_profile_sensor_data)
        self.odom_subscriber_ = self.create_subscription(Odometry, 'odom', self.odom_callback, qos_profile)
        self.timer_ = self.create_timer(1.0, self.timer_callback)

    def odom_callback(self, msg):
        self.get_logger().info(f'Received odom: {msg}')
        # self.odom_publisher_.publish(msg)

    def timer_callback(self):
        odom_msg = Odometry()
        self.get_logger().info('Publishing odom')
        self.odom_publisher_.publish(odom_msg)

def main(args=None):
    rclpy.init(args=args)
    node = OdomPublisherSubscriber()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

File: ch10/ch10_ws/src/learn_qos_cpp/src/reliability_test.cpp
#include <nav_msgs/msg/odometry.hpp>
#include <rclcpp/rclcpp.hpp>

class OdomPublisherSubscriber : public rclcpp::Node
{
public:
    OdomPublisherSubscriber() : Node("odom_publisher_subscriber")
    {
        rclcpp::QoS qos_profile(10);                                       // Queue depth of 10
        qos_profile.reliability(RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT);   // Reliability policy
        qos_profile.durability(RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL); // Durability policy
        qos_profile.history(RMW_QOS_POLICY_HISTORY_KEEP_LAST);             // History policy
        qos_profile.deadline(rclcpp::Duration(1, 0));                      // Deadline of 1 second

        odom_publisher_ = this->create_publisher<nav_msgs::msg::Odometry>(
            "odom", qos_profile);

        odom_subscription_ = this->create_subscription<nav_msgs::msg::Odometry>(
            "odom", rclcpp::SensorDataQoS(),
            [this](const nav_msgs::msg::Odometry::SharedPtr msg)
            {
                (void)msg;
                RCLCPP_INFO(this->get_logger(), "Received odometry message");
            });

        // Create a 1-second timer and specify the callback function
        timer_ = this->create_wall_timer(std::chrono::seconds(1), [this]()
                                         { odom_publisher_->publish(nav_msgs::msg::Odometry()); });
    }

private:
    rclcpp::Publisher<nav_msgs::msg::Odometry>::SharedPtr odom_publisher_;
    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_subscription_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto odom_node = std::make_shared<OdomPublisherSubscriber>();
    rclcpp::spin(odom_node);
    rclcpp::shutdown();
    return 0;
}